core/storage/blog_post/gae_models.py

```diff
@@ -0,0 +1,457 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for storing data related to blog posts."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.platform import models
+import python_utils
+import utils
+(base_models, user_models) = models.Registry.import_models([
+    models.NAMES.base_model, models.NAMES.user])
+
+datastore_services = models.Registry.import_datastore_services()
+
+
+class BlogPostModel(base_models.BaseModel):
+    """Model to store blog post data.
+
+    The id of instances of this class is in the form of random hash of
+    12 chars.
+    """
+
+    # We use the model id as a key in the Takeout dict.
+    ID_IS_USED_AS_TAKEOUT_KEY = True
+
+    # The ID of the user the blog post is authored by.
+    author_id = datastore_services.StringProperty(required=True)
+    # Title of the blog post.
+    title = datastore_services.StringProperty(indexed=True, required=True)
+    # Content of the blog post.
+    content = datastore_services.TextProperty(indexed=False, required=True)
+    # The unique url fragment of the blog post. If the user directly enters the
+    # blog post's url in the editor or the homepage, the blogPostModel will be
+    # queried using the url fragment to retrieve data for populating the editor
+    # dashboard / blog post page.
+    url_fragment = (
+        datastore_services.StringProperty(indexed=True, required=True))
+    # Tags associated with the blog post. 'repeated' property is set to true
+    # to take a list of values of underlying type(str).
+    tags = datastore_services.StringProperty(indexed=True, repeated=True)
+    # The thumbnail filename of the blog post.
+    thumbnail_filename = datastore_services.StringProperty(required=True)
+    # Time when the blog post model was published.
```

Comment:
Seems like it's not required, so explain in this docstring what a value of None means.

Also a blog post may be published multiple times, right? (if it's published then unpublished then published again). Maybe clarify that this refers to the time when it was last published.

Reply:
Added doc string -
    ' Time when the blog post model was last published. Value will be None if the blog post is a draft i.e it is not published.'

Reply:
Looks good, but just to check, can blog posts be unpublished after they're published? If so, then this is slightly inaccurate because the value might be non-None even if the blog post is not currently published. That can be fixed by changing the last part to "if the blog post has never been published."

Reply:
Fixed the docstring as it can be unpublished and then published.